internal/policygen/iam.go

```diff
@@ -15,62 +15,103 @@
-	// TODO(https://github.com/GoogleCloudPlatform/healthcare-data-protection-suite/issues/152): Generate policies.
 
+	for root, rbs := range bindings {
+		outputFolder := fmt.Sprintf("%s_%s", root.Type, root.ID)
+		// Generate policies for allowed roles.
+		data := map[string]interface{}{
+			"target": fmt.Sprintf("%s/%s", root.Type, root.ID),
+			"roles":  rbs,
+		}
+		in := filepath.Join(templateDir, "forseti", "tf_based", "iam_allow_roles.yaml")
+		out := filepath.Join(outputPath, forsetiOutputRoot, outputFolder, "iam_allow_roles.yaml")
+		if err := template.WriteFile(in, out, data); err != nil {
+			return err
+		}
+
+		// Generate policies for allowed bindings for each role.
+		for role, bindings := range rbs {
+			data := map[string]interface{}{
+				"target":   fmt.Sprintf("%s/%s", root.Type, root.ID),
+				"role":     role,
+				"bindings": bindings,
+			}
+			in := filepath.Join(templateDir, "forseti", "tf_based", "iam_allow_bindings.yaml")
+			out := filepath.Join(outputPath, forsetiOutputRoot, outputFolder, fmt.Sprintf("iam_allow_bindings_%s.yaml", strings.TrimPrefix(role, "roles/")))
+			if err := template.WriteFile(in, out, data); err != nil {
+				return err
+			}
+		}
+	}
 	return nil
 }
 
-func bindings(resources []*states.Resource) ([]*binding, error) {
+func allBindings(rn runner.Runner, resources []*states.Resource) error {
 	typeToIDField := map[string]string{
 		"project":      "project",
 		"folder":       "folder",
 		"organization": "org_id",
 	}
 
-	var bindings []*binding
 	for t, idField := range typeToIDField {
+		// TODO(https://github.com/GoogleCloudPlatform/healthcare-data-protection-suite/issues/152): Support google_*_iam_bindings.
 		resourceType := fmt.Sprintf("google_%s_iam_member", t)
 		instances, err := terraform.GetInstancesForType(resources, resourceType)
 		if err != nil {
-			return nil, fmt.Errorf("get resource instances for type %q: %v", resourceType, err)
+			return fmt.Errorf("get resource instances for type %q: %v", resourceType, err)
 		}
 
 		for _, ins := range instances {
 			if err := validate(ins, []string{idField, "role", "member"}); err != nil {
-				return nil, err
+				return err
+			}
+
+			id := ins[idField].(string) // Type checked in validate()
+			// For projects, the ID in the state is the project ID, but we need project number in policies.
+			if t == "project" {
+				if id, err = projectNumber(rn, id); err != nil {
+					return err
+				}
+			}
+
+			key := root{Type: t, ID: id}
+
+			// Init the roleBindings map if it didn't exist.
+			if _, ok := bindings[key]; !ok {
+				bindings[key] = make(roleBindings)
 			}
 
-			bindings = append(bindings, &binding{
-				Type:   t,
-				ID:     ins[idField].(string), // Type checked in validate()
-				Role:   ins["role"].(string),
-				Member: ins["member"].(string),
-			})
+			bindings[key][ins["role"].(string)] = append(bindings[key][ins["role"].(string)], ins["member"].(string))
```

Comment:
`ins["role"].(string)` is repeated, should be extracted. Or just use `id`.

Reply:
Done